lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -432,6 +421,19 @@ def _is_valid_rule(rule: dict, allow_free_standing: bool) -> bool:
     return True
 
 
+def _type_convert_stored(obj):
+    """Convert Stored* to their appropriate types, recursively."""
+    if isinstance(obj, StoredList):
+        return list(map(_type_convert_stored, obj))
+    elif isinstance(obj, StoredDict):
+        rdict = {}  # type: Dict[Any, Any]
+        for k in obj.keys():
+            rdict[k] = _type_convert_stored(obj[k])
+        return rdict
+    else:
+        return obj
+
+
```

Comment:
These conversion operations seems like something we probably should bug the operator framework team about providing as part of the Stored* datatypes down the line. Wdyt, @jnsgruk?

Reply:
This was discussed at length -- and rejected. Definitely a candidate for `observability-libs`, though.

Reply:
(context: https://github.com/canonical/operator/pull/572)

Reply:
saying that, the PR I link about was about implementing magic on the Stored* types. Perhaps a more explicit conversion might be more palatable.

Reply:
I feel `StoredDict` should have a `.to_dict` like we do for Pebble services and plans, etc.

Reply:
That's kind of a can of worms.

The initial impetus behind that OF PR came about early in the charming campaign, when Balbir quickly realized that YAML/JSON dumping didn't work on `Stored*` objects, because they don't behave like the types they're named after (in a rather large variety of ways, but in that way particularly).

The PR was principally about hiding this from users by implementing equality methods with transparent conversions, and letting them be converted to JSON/YAML without the use of an explicit YAML/JSON formatter. My feeling then (and now) is that it doesn't feel Pythonic to have `StoredDict.as_dict()`, or to need a formatter for YAML/JSON, since a `FooDict` and `FooList` should just behave like the types they are claiming to be.

In the end, we had a discussion about whether `StoredState` (and the objects underneath it) are/were even appropriate places to keep data. You can't put the cat back in the bag, but certainly "we" at Canonical can write charms in a way which reflects best practice. For resiliency, as your recent comments about `StoredState` being harmful allude to, we may be better off creating peer relation buckets as the default and keeping data there, which is battle-tested, has a well-known API (anything that can be converted to/from a string), etc.

I'd suggest that, instead of going the route of explicit conversions (which will actually be `_type_convert_stored` anyway), a peer relation bucket is added and used here instead. The code block in the middle of [this comment](https://github.com/canonical/operator/pull/572#issuecomment-931529580) trivially demonstrates why. The existing magic conversions in OF already turn a list under a dict into a `StoredList`. Every node under the parent becomes a `Stored*` type, and they all need to be converted back recursively if it needs to work with `json.dumps()` or `yaml.dump()`, so an explicit conversion of `StoredDict.to_dict()` is necessarily more or less identical to `_type_convert_stored`, and the same goes for `StoredList.to_list()`.

Let's add and use a peer relation bucket, which gives anyone reading the code for this when writing a new charm a sense of "best practice" without list of footnotes illustrating the footguns from `StoredState`

Reply:
I agree with you @rbarry82, but let's do that in a separate PR/issue 😅

Reply:
Let's follow this in #50